src/algebra/monoid_algebra/no_zero_divisors.lean

```diff
@@ -152,4 +184,93 @@ end⟩
 
 end left_or_right_orderability
 
+section covariant_lt
+variables [has_add A] [partial_order A] [covariant_class A A (+) (<)] {a t b : A} {f g : A →₀ R}
+
+/--  The "top" element of `finsupp.single a r * f`  is the product of `r` and
+the "top" element of `f`.  Here, "top" is simply an upper bound for the elements
+of the support of `f` (e.g. the product is `0` if "top" is not a maximum).
+
+The corresponding statement for a general product `f * g` is `add_monoid_algebra.mul_apply_of_le`.
+It is proved with a further `covariant_class` assumption. -/
+lemma single_mul_apply_of_le (r : R) (ft : ∀ a ∈ f.support, a ≤ t) :
+  (finsupp.single a r * f : add_monoid_algebra R A) (a + t) = r * f t :=
+begin
+  classical,
+  nth_rewrite 0 ← f.erase_add_single t,
+  rw [mul_add, single_mul_single, finsupp.add_apply, finsupp.single_eq_same],
+  convert zero_add _,
+  refine finsupp.not_mem_support_iff.mp (λ h, _),
+  refine not_not.mpr ((support_mul (finsupp.single a r) (f.erase t)) h) _,
+  simp only [finsupp.mem_support_single, finset.mem_bUnion, ne.def, finset.mem_singleton,
+    exists_prop, not_exists, not_and, and_imp, forall_eq, finsupp.support_erase],
+  intros _ x xs,
+  refine (add_lt_add_left (with_bot.coe_lt_coe.mp _) _).ne',
+  refine with_bot.coe_lt_coe.mpr ((ft _ (finset.mem_of_mem_erase xs)).lt_of_ne _),
+  exact finset.ne_of_mem_erase xs,
+end
```

Comment:
Here's an alternate proof that I found slightly easier to understand:
```lean
begin
  classical,
  rw [mul_apply, finsupp.sum_single_index],
  { simp only [←mul_zero r] { single_pass := tt },
    simp_rw [←mul_ite, finsupp.sum, ←finset.mul_sum, ←pi.single_apply (a + t)],
    congr' 1,
    refine (finset.sum_eq_single t _ _).trans (pi.single_eq_same _ _),
    { intros b hb hbt,
      have := (add_lt_add_left ((ft _ hb).lt_of_ne hbt) a).ne,
      exact pi.single_eq_of_ne this _ },
    { intro h,
      rw pi.single_eq_same,
      exact not_mem_support_iff.mp h } },
  { simp_rw [finsupp.sum, zero_mul, if_t_t, finset.sum_const_zero] },
end
```
The crux seems to be that `ft` combined with the order hypothesis says "addition in cancellative in the right region on the support"